pkg/aspect/manager.go

```diff
@@ -28,50 +29,96 @@ import (
 	"istio.io/mixer/pkg/expr"
-	"istio.io/mixer/pkg/status"
 )
 
 type (
-	// Output captures the output from invoking an aspect.
-	Output struct {
-		// status code
-		Status rpc.Status
-
-		// Additional method-specific returned state
-		Response APIMethodResp
-
-		//TODO attribute mutator
-		//If any attributes should change in the context for the next call
-		//context remains immutable during the call
-	}
-
 	// Manager is responsible for a specific aspect and presents a uniform interface
 	// to the rest of the system.
 	Manager interface {
 		config.AspectValidator
 
-		// NewAspect creates a new aspect instance given configuration.
-		NewAspect(cfg *cpb.Combined, adapter adapter.Builder, env adapter.Env, df descriptor.Finder) (Wrapper, error)
-
-		// Kind return the kind of aspect
+		// Kind return the kind of aspect handled by this manager
 		Kind() Kind
 	}
 
-	// Wrapper encapsulates a single aspect and allows it to be invoked.
-	Wrapper interface {
+	// CheckManager take care of aspects used to implement the Check API method
+	CheckManager interface {
+		Manager
+
+		// NewCheckExecutor creates a new aspect executor given configuration.
+		NewCheckExecutor(cfg *cpb.Combined, builder adapter.Builder, env adapter.Env, df descriptor.Finder) (CheckExecutor, error)
+	}
+
+	// ReportManager take care of aspects used to implement the Report API method
+	ReportManager interface {
+		Manager
+
+		// NewReportExecutor creates a new aspect executor given configuration.
+		NewReportExecutor(cfg *cpb.Combined, builder adapter.Builder, env adapter.Env, df descriptor.Finder) (ReportExecutor, error)
+	}
+
+	// QuotaManager take care of aspects used to implement the Quota API method
+	QuotaManager interface {
+		Manager
+
+		// NewQuotaExecutor creates a new aspect executor given configuration.
+		NewQuotaExecutor(cfg *cpb.Combined, builder adapter.Builder, env adapter.Env, df descriptor.Finder) (QuotaExecutor, error)
+	}
+
+	// Executor encapsulates a single aspect and allows it to be invoked.
+	Executor interface {
 		io.Closer
+	}
 
-		// Execute dispatches to the adapter.
-		Execute(attrs attribute.Bag, mapper expr.Evaluator, ma APIMethodArgs) Output
+	// CheckExecutor encapsulates a single CheckManager aspect and allows it to be invoked.
+	CheckExecutor interface {
+		Executor
+
+		// Execute dispatches to the aspect manager.
+		Execute(attrs attribute.Bag, mapper expr.Evaluator) rpc.Status
+	}
+
+	// ReportExecutor encapsulates a single ReportManager aspect and allows it to be invoked.
+	ReportExecutor interface {
+		Executor
+
+		// Execute dispatches to the aspect manager.
+		Execute(attrs attribute.Bag, mapper expr.Evaluator) rpc.Status
+	}
+
+	// QuotaExecutor encapsulates a single QuotaManager aspect and allows it to be invoked.
+	QuotaExecutor interface {
+		Executor
+
+		// Execute dispatches to the aspect manager.
+		Execute(attrs attribute.Bag, mapper expr.Evaluator, qma *QuotaMethodArgs) (rpc.Status, *QuotaMethodResp)
```

Comment:
Small nit: I think it's more idiomatic to flip around these return params: even though it's not directly `error`, `rpc.Status` is how we communicate the success or failure of the call, and QuotaMethodResp is the primary return value of the call. Looking through the quota manager impl, it's strikes me as a little odd that in error cases we're doing `return foo, nil` rather than the reverse.